Faster than Real -Time support in GZ

dirksavage88 opened the pull request:
<!--

Thank you for your contribution!

Get early feedback through
- Dronecode Discord: https://discord.gg/dronecode
- PX4 Discuss: http://discuss.px4.io/
- opening a draft pr and sharing the link

-->

### Solved Problem
The new GZ simulation real time factor is configurable, however the service request has not been implemented in the GZ bridge. 

partially fixes Fixes https://github.com/PX4/PX4-Autopilot/issues/23602

### Solution
- Configure the GZ bridge to send a service request with the simulation speedup value environment variable "PX4_SIM_SPEED_FACTOR"

### Changelog Entry
For release notes:
```Implement configurable real time factor in GZ sim (PX4_SIM_SPEED_FACTOR)
Feature: Configurable real time factor in new GZ sim
New parameter: None
Documentation: Configurable via environment variable "PX4_SIM_SPEED_FACTOR"
```

### Alternatives
Change the real time facto in gazebo world sdf manually, or make a world file specific to an increased RTF

### Test coverage
- Tested in GZ sim
[rtf.webm](https://github.com/user-attachments/assets/7010c3d6-5d51-437b-bace-0a8eb4626040)

### Context
Related links, screenshot before/after, video

mrpollo commented:
@Jaeyoung-Lim @dagar can you please give this a review? seems like a good addition

Jaeyoung-Lim commented on src/modules/simulation/gz_bridge/GZBridge.cpp:
```diff
@@ -211,6 +211,23 @@ int GZBridge::init()
+		if (!callPhysicsMsgService(physics_service, p_req)) {
+			return PX4_ERROR;
+		}
+	}
```
It is probably harmless, but should we maybe configure the physics before spawning the model?

dirksavage88 replied:
I placed the physics service call further up in the init method, but not sure if this is before the model spawn you mentioned @Jaeyoung-Lim

Jaeyoung-Lim reviewed (changes requested):
Amazing thanks!

- Just one small nitpick on the order of configuring the physics.

- More of a question/note: Since we configure the physics in the gz bridge, this might cause problems when running multiple px4 instances. Might be worth taking it out to somewhere else in the future.

hamishwillee commented:
This will good and will need docs hint hint

dirksavage88 commented:
@Jaeyoung-Lim is there anything preventing this from going in? I've addressed the comments so far. I think @hamishwillee is right that we will need a docs update after it goes in

Jaeyoung-Lim commented:
@dirksavage88 As I commented above, you are configuring the physics only when a model exists.

However, physics is more of a world parameter and therefore should not be related to the model itself, but rather the world.

Therefore I think it would be better if the config is done outside of gz_bridge.

dirksavage88 commented:
@Jaeyoung-Lim I'm going push back and say this is a nitpick. What is the functional benefit to placing the physics outside of the gz_bridge? 

Also the world services are created in gz_bridge here: https://github.com/PX4/PX4-Autopilot/blob/04a3c4af20df9e851e281776bfa51d02759ddee6/src/modules/simulation/gz_bridge/GZBridge.cpp#L117

Should not the physics also be configured along with the world services?

Jaeyoung-Lim commented:
@dirksavage88 Yes, and that needs to change if we want to support multiple vehicles.

dirksavage88 commented:
@Jaeyoung-Lim okay that is making sense now. So there should be somewhere else in the code where we can do both world services and physics.

Jaeyoung-Lim commented:
@dirksavage88 Probably rethinking this - lets get this in as is, since I dont want to slow you down due to architectural improvements.

But lets put a note somewhere that we need to fix this, so that it flows nicely to multi vehicle support.

mrpollo commented:
Also please create an issue to track this, make sure to assign yourself.

dakejahl commented:
> this would break if there are multiple vehicles in the simulation with the gz_bridge

So this only breaks if each px4-sitl instance has a different value for PX4_SIM_SPEED_FACTOR. Which is probably pretty easy to screw up. I don't think there's a way to set this directly in the world SDF though. I agree we should get it in now and fix later.

hamishwillee commented:
@dirksavage88 @Jaeyoung-Lim Re docs, is this as simple as updating https://docs.px4.io/main/en/simulation/#run-simulation-faster-than-realtime to mention Gazebo (GZ) supports this too?

Are there any specific things to think about that are different on GZ?

dirksavage88 commented:
> @dirksavage88 @Jaeyoung-Lim Re docs, is this as simple as updating https://docs.px4.io/main/en/simulation/#run-simulation-faster-than-realtime to mention Gazebo (GZ) supports this too?
> 
> Are there any specific things to think about that are different on GZ?

I think the COM_DL_LOSS_T param still applies, however I do notice that in the new gz with a speed up factor of 10x, I am not able to get local/global position to work. I am not sure if this was also the case for gz classic. I can get the x500 to arm in SITL with a speed up of 5x though

dakejahl commented:
> however I do notice that in the new gz with a speed up factor of 10x, I am not able to get local/global position to work

@dirksavage88 the issue is that this PR increases the time step rather than the real_time_factor. I have an open PR which is a large refactor of the the GZBridge and px4-rc.simulator which fixes this. If you want to try it out and leave a review!
https://github.com/PX4/PX4-Autopilot/pull/24421

hamishwillee commented:
Thanks very much.
Guys, when you merge something, please try update the docs. The chances of me missing stuff like this in PRs is high. I won't miss a PR.

Needs review: https://github.com/PX4/PX4-user_guide/pull/3628
